multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -90,57 +93,113 @@ func NewMemberClusterSetReconciler(client client.Client,
+			// Leader cluster will delete the stale MemberClusterAnnounce with a garbage collection mechanism in this case.
+			return fmt.Errorf("failed to delete MemberClusterAnnounce in the leader cluster: %v", err)
+		}
+		r.remoteCommonArea.Stop()
+		r.remoteCommonArea = nil
+		r.clusterSetConfig = nil
+		if isNotFound {
+			r.clusterID = common.InvalidClusterID
+			r.clusterSetID = common.InvalidClusterSetID
+		}
+		klog.InfoS("Clean up all resources created by Antrea Multi-cluster Controller", "clusterset", req.NamespacedName)
+		if err := cleanUpResourcesCreatedByMC(ctx, r.Client); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	r.commonAreaLock.Lock()
+	if clusterSetNotFound {
 		klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
 		if r.remoteCommonArea != nil {
-			if err := r.deleteMemberClusterAnnounce(ctx); err != nil {
-				// MemberClusterAnnounce could be kept in the leader cluster, if antrea-mc-controller crashes after the failure.
-				// Leader cluster will delete the stale MemberClusterAnnounce with a garbage collection mechanism in this case.
-				return ctrl.Result{}, fmt.Errorf("failed to delete MemberClusterAnnounce in the leader cluster: %v", err)
+			if err := cleanUpResources(clusterSetNotFound); err != nil {
+				r.commonAreaLock.Unlock()
+				return ctrl.Result{}, err
 			}
-			r.remoteCommonArea.Stop()
-			r.remoteCommonArea = nil
-			r.clusterSetConfig = nil
-			r.clusterID = common.InvalidClusterID
-			r.clusterSetID = common.InvalidClusterSetID
 		}
+		r.commonAreaLock.Unlock()
 		return ctrl.Result{}, nil
 	}
 
 	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
 
 	// Handle create or update
-	if r.clusterSetConfig == nil {
-		r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		r.clusterSetID = common.ClusterSetID(clusterSet.Name)
-		if clusterSet.Spec.ClusterID == "" {
-			// ClusterID is a required feild, and the empty value case should only happen
-			// when Antrea Multi-cluster is upgraded from an old version prior to v1.13.
-			// Here we try to update the ClusterSet's ClusterID when it's configured in an
-			// existing ClusterClaim.
-			clusterSet.Spec.ClusterID = string(r.clusterID)
-			err = r.Update(context.TODO(), clusterSet)
-			if err != nil {
-				klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
+
+	newLeader := clusterSet.Spec.Leaders[0]
+	if r.installedLeader.clusterID == newLeader.ClusterID && r.installedLeader.serverUrl == newLeader.Server &&
+		r.installedLeader.secretName == newLeader.Secret {
+		klog.V(2).InfoS("No change for leader cluster configuration")
+		r.commonAreaLock.Unlock()
+		return ctrl.Result{}, nil
+	}
+
+	var clusterSetCreated bool
+	if r.clusterID == "" || r.clusterID == common.InvalidClusterID {
```

Comment:
Could not we set r.clusterID to `common.InvalidClusterID` at initialization, so we need not to check both?

Reply:
I feel this check is not right. We should check if ClusterID/ClusterSet ID changes, not if clusterID is set, as before cleanup succeeds you do not set `r.clusterID` to nil.

Reply:
Yeah, set the r.clusterID to common.InvalidClusterID at initialization.